radicle/src/ssh.rs

```diff
@@ -75,30 +78,81 @@ pub mod fmt {
     }
 }
 
+#[derive(Debug, Error)]
+pub enum SignatureError {
+    #[error(transparent)]
+    Invalid(#[from] crypto::Error),
+    #[error(transparent)]
+    Encoding(#[from] encoding::Error),
+    #[error("unknown algorithm '{0}'")]
+    UnknownAlgorithm(String),
+}
+
+impl Encodable for crypto::Signature {
+    type Error = SignatureError;
+
+    fn read_ssh(r: &mut encoding::Cursor) -> Result<Self, Self::Error> {
+        let buf = r.read_string()?;
+        let mut r2 = buf.reader(0);
+
+        let sig_type = r2.read_string()?;
+        if sig_type != b"ssh-ed25519" {
+            return Err(SignatureError::UnknownAlgorithm(
+                String::from_utf8_lossy(sig_type).to_string(),
+            ));
+        }
+        let sig = crypto::Signature::try_from(r2.read_string()?)?;
+
+        Ok(sig)
+    }
+
+    fn write_ssh<E: Encoding>(&self, buf: &mut E) {
+        let mut buf2 = Vec::new();
+        buf2.extend_ssh_string(b"ssh-ed25519");
+        buf2.extend_ssh_string(self.as_ref());
+        buf.extend_ssh_string(&buf2);
+    }
+}
+
 #[derive(Debug, Error)]
 pub enum PublicKeyError {
     #[error(transparent)]
     Invalid(#[from] crypto::Error),
     #[error(transparent)]
     Encoding(#[from] encoding::Error),
+    #[error("unknown algorithm '{0}'")]
+    UnknownAlgorithm(String),
 }
 
-impl radicle_ssh::key::Public for PublicKey {
+impl Encodable for PublicKey {
     type Error = PublicKeyError;
 
-    fn write(&self, buf: &mut Zeroizing<Vec<u8>>) -> usize {
-        let mut n = 0;
-        let typ = b"ssh-ed25519";
-        let size = typ.len() + self.len() + mem::size_of::<u32>() * 2;
+    fn read_ssh(r: &mut encoding::Cursor) -> Result<Self, Self::Error> {
+        let buf = r.read_string()?;
+        let mut str_r = buf.reader(0);
+        match str_r.read_string()? {
+            b"ssh-ed25519" => {
+                let s = str_r.read_string()?;
+                let p = PublicKey::try_from(s)?;
 
-        buf.write_u32::<BigEndian>(size as u32)
-            .expect("writing to a vector never fails");
+                Ok(p)
+            }
+            v => Err(PublicKeyError::UnknownAlgorithm(
+                String::from_utf8_lossy(v).to_string(),
+            )),
+        }
+    }
 
-        n += mem::size_of::<u32>(); // The blob size.
-        n += buf.extend_ssh_string(typ);
-        n += buf.extend_ssh_string(&self[..]);
-        n
+    fn write_ssh<E: Encoding>(&self, w: &mut E) {
+        let mut str_w: Vec<u8> = Vec::<u8>::new();
+        str_w.extend_ssh_string(b"ssh-ed25519");
+        str_w.extend_ssh_string(&self[..]);
+        w.extend_ssh_string(&str_w);
     }
```

Comment:
This is basically what `PublicKey::write` does just below

Reply:
I'm hoping to replace the SecretKey/PublicKey traits with the Encodable trait in upcoming changes.